akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -3,31 +3,95 @@
  */
 package akka.stream.scaladsl
 
-import java.io.{ BufferedWriter, File, FileWriter, Writer }
-import java.util.concurrent.Executors
-
 import akka.Done
 import akka.stream.ActorAttributes._
 import akka.stream.Supervision._
-import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
 import akka.stream.testkit.StreamSpec
 import akka.stream.testkit.Utils._
 import akka.stream.testkit.scaladsl.TestSource
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings }
 import org.scalatest.BeforeAndAfter
 
-import scala.concurrent.{ Await, ExecutionContext, Future, Promise }
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{ Await, Future, Promise }
 
 class FoldResourceAsyncSinkSpec extends StreamSpec(UnboundedMailboxConfig) with BeforeAndAfter {
-  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
-  implicit val materializer = ActorMaterializer(settings)
 
-  val ioExecutor = Executors.newFixedThreadPool(4)
-  val ioExecutionContext = ExecutionContext.fromExecutor(ioExecutor)
+  class TestResource[T] {
+
+    var autoCompleteOpen: Boolean = true
+    var autoCompleteWrite: Boolean = true
+    var autoCompleteClose: Boolean = true
+
+    var isOpened = false
+    var isClosed = false
+    var written: Vector[T] = _
+    var completeOpen: Promise[Unit] = _
+    var completeWrite: Promise[Unit] = _
+    var completeClose: Promise[Unit] = _
+
+    def open(): Future[TestResource[T]] = {
+      def syncOpen(): TestResource[T] = {
+        isOpened = true
+        isClosed = false
+        // written is cleared on open
+        written = Vector.empty[T]
+        this
+      }
+
+      if (autoCompleteOpen) {
+        Future.successful(syncOpen())
+      } else {
+        completeOpen = Promise[Unit]
+        completeOpen.future.map { _ ⇒
+          syncOpen()
+        }
+      }
+    }
+
+    def write(t: T): Future[Unit] = {
+      def syncWrite(): Unit = {
+        if (!isOpened) {
+          sys.error("Writing to a not opened resource")
+        }
+        if (isClosed) {
+          sys.error("Writing to a closed resource")
+        }
+        written = written :+ t
+      }
+
+      if (autoCompleteWrite) {
+        Future.successful(syncWrite())
+      } else {
+        completeWrite = Promise[Unit]
+        completeWrite.future.map { _ ⇒
+          syncWrite()
+        }
+      }
+    }
+
+    def close(): Future[Unit] = {
+      def syncClose(): Unit = {
+        if (!isOpened) {
+          sys.error("Closing a not opened resource")
+        }
+        isClosed = true
+      }
+      if (autoCompleteClose) {
+        Future.successful(syncClose())
+      } else {
+        completeClose = Promise[Unit]
+        completeClose.future.map { _ ⇒
+          syncClose()
+        }
+      }
+    }
```

Comment:
The fields that are read from the test thread needs to be `@volatile` (or threadsafed in some other way) or else there is no guarantee the assertions will see the changes that happened while executing the stream (which happens on a different thread)